src/Swarm/Game/ScenarioStatus.hs

```diff
@@ -0,0 +1,319 @@
+  _NotStarted,
+  _InProgress,
+  _Complete,
+  ScenarioInfo,
+  scenarioPath,
+  scenarioStatus,
+  scenarioBest,
+  updateScenarioInfoOnQuit,
+
+  -- * Scenario collection
+  ScenarioCollection (..),
+  scenarioCollectionToList,
+  scenarioItemByPath,
+  normalizeScenarioPath,
+  ScenarioItem (..),
+  scenarioItemName,
+  _SISingle,
+
+  -- * Loading and saving scenarios
+  loadScenarios,
+  loadScenarioInfo,
+  saveScenarioInfo,
+
+  -- * Re-exports
+  module Swarm.Game.Scenario,
+) where
+
+import Control.Algebra (Has)
+import Control.Carrier.Lift (Lift, sendIO)
+import Control.Carrier.Throw.Either (Throw, runThrow, throwError)
+import Control.Lens hiding (from, (<.>))
+import Control.Monad (unless, when)
+import Data.Aeson (
+  Options (..),
+  defaultOptions,
+  genericParseJSON,
+  genericToEncoding,
+  genericToJSON,
+ )
+import Data.Char (isSpace, toLower)
+import Data.Function (on)
+import Data.List (intercalate, stripPrefix, (\\))
+import Data.Map (Map)
+import Data.Map qualified as M
+import Data.Maybe (isJust)
+import Data.Text (Text, pack)
+import Data.Time (NominalDiffTime, ZonedTime, diffUTCTime, zonedTimeToUTC)
+import Data.Yaml as Y
+import GHC.Generics (Generic)
+import Paths_swarm (getDataDir)
+import Swarm.Game.Entity
+import Swarm.Game.Scenario
+import Swarm.Util (getSwarmSavePath)
+import System.Directory (canonicalizePath, doesDirectoryExist, doesFileExist, listDirectory)
+import System.FilePath (pathSeparator, splitDirectories, takeBaseName, takeExtensions, (-<.>), (</>))
+import Witch (into)
+
+instance Eq ZonedTime where
+  (==) = (==) `on` zonedTimeToUTC
+
+instance Ord ZonedTime where
+  (<=) = (<=) `on` zonedTimeToUTC
+
+data ScenarioStatus
+  = NotStarted
+  | InProgress
+      { -- | Time when the scenario was started including time zone.
+        _scenarioStarted :: ZonedTime
+      , -- | Time elapsed until quitting the scenario.
+        _scenarioElapsed :: NominalDiffTime
+      }
+  | Complete
+      { -- | Time when the scenario was started including time zone.
+        _scenarioStarted :: ZonedTime
+      , -- | Time elapsed until quitting the scenario.
+        _scenarioElapsed :: NominalDiffTime
+      }
+  deriving (Eq, Ord, Show, Read, Generic)
+
+instance FromJSON ScenarioStatus where
+  parseJSON = genericParseJSON scenarioOptions
+
+instance ToJSON ScenarioStatus where
+  toEncoding = genericToEncoding scenarioOptions
+  toJSON = genericToJSON scenarioOptions
+
+data ScenarioInfo = ScenarioInfo
+  { _scenarioPath :: FilePath
+  , _scenarioStatus :: ScenarioStatus
+  , _scenarioBest :: ScenarioStatus
+  }
+  deriving (Eq, Ord, Show, Read, Generic)
+
+instance FromJSON ScenarioInfo where
+  parseJSON = genericParseJSON scenarioOptions
+
+instance ToJSON ScenarioInfo where
+  toEncoding = genericToEncoding scenarioOptions
+  toJSON = genericToJSON scenarioOptions
+
+scenarioOptions :: Options
+scenarioOptions =
+  defaultOptions
+    { fieldLabelModifier = map toLower . drop (length "_scenario")
```

Comment:
Cool, didn't know about this!

Reply:
It was mentioned in the SO answer you linked in #587. :wink:

There are more [`Options`](https://hackage.haskell.org/package/aeson-2.1.0.0/docs/Data-Aeson.html#t:Options), and they make the YAML more readable, so I'll tweak the other instances in the future too.